SWIPs/swip-honey_to_money.md

```diff
@@ -7,51 +7,66 @@ Type: Standards track
+* A price oracle returns a `honeyPrice` object, specifying a `Time To Live (TTL)` in seconds and an array of `price` objects, each index specifying a `validFrom` filed as an UTCDate and a `honeyPrice`. To summarize, the oracle returns:
+```typescript 
+honeyPrices: { 
+ TTL,
+ prices: [0: {validFrom, honeyPrice}, n: {validFrom, honeyPrice}]
+}
+```
+* The applied price is the `price` where `validFrom` is in the most recent past. 
+* Upon receipt of a payment or intent to send a payment, nodes will check their local cache on the existence of a non-expired `honeyPrice` object. If this exists, the oracle will not be queried, if it does not exist, the oracle will be queried and the answer stored in the local cache of the node for `TTL` time. 
+* The number of payments received/sent, `TTL` of `honeyPrices`, as well as the costs to query the oracle (computational overload/time delay), should determine whether nodes actively keep their local cache up-to-date with `honeyPrices` or whether they will passively request a new `honeyPrices` object when needed.
 
 ## Rationale
-We need a way for prices in the Swarm to change over time. However, due to the nature of Swarm - specifically independent connections between pairs of nodes each with their individual accounting and payments - there is no clear price discovery mechanism. Furthermore, since Swarm nodes are expected to be conduits for both payments and data (buy from one peer and sell to the other) it is necessary for nodes to coordinate price changes with all their peers. In short, we need a mechanism for network-wide changes in pricing that the peer-to-peer routing and accounting does not provide.
-By externalising the price to an oracle, we can completely decouple Swarm's core functions from price discovery. By building support for price oracles, we can coordinate price changes across the network, and set the stage for future experimentation with various mechanisms for setting prices, from DAOs, to bonding curves, to fixed prices. This SWIP makes no recommendation for the eventual pricing mechanism, it only prepares the network for adopting them in future.
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+### General rationale for choosing oracles
+There exist a need for the price of Swarm to change over time. However, due to the nature of Swarm, there is a need for this price to be decided network-wide, instead of peer-to-peer. Specifically:
+- The requirement on connectivity in the local bin and far-away bins means that nodes *must agree* with all nodes in their local bin and at least one node in a far-away bin on the price.
```

Comment:
why is this the case?

Reply:
Based on [Swarm-Docs/architecture](https://swarm-guide.readthedocs.io/en/latest/architecture.html)
>**2.2.2 Kademlia Topology**
We say that a node has kademlia connectivity if (1) it is connected to at least one node for each proximity order up to (but excluding) some maximum value d (called the saturation depth) and (2) it is connected to all nodes whose proximity order relative to the node is greater or equal to d

Reply:
alright, that makes sense. but why is kademlia connectivity related to an agreement on price? why _must_ the nodes agree on this? 

are we saying that they should? if so, why?

Reply:
I was wondering about this sentence too (for a different reason) and I think now that it should be amended. This sentence suggests that all nodes in the kademlia bins need to agree on the price *a priori* in order to do business, which is not what we are doing now. Currently, connectivity is independent from price. Kademlia is built up with devp2p discovery, and price only is an issue when we exchange cheques. So kademlia is only indirectly affected, in fact only at the point that if cheques are exchanged and one peer does not agree with the price of the other, a disconnection occurs and the peer is kicked out of the kademlia.

Even if the multi-payment SWIP would be implemented, the handshake only includes the currency type and their contracts, not the actual price, so this sentence IMHO gives a wrong message, so it should be removed.

Reply:
> This sentence suggests that all nodes in the kademlia bins need to agree on the price a priori in order to do business, which is not what we are doing 

This is indeed not what we are doing now and this is also not what is described here. The sentence gives an argument for why prices must be decided network-wide instead of peer-to peer. I would feel bad about removing this sentence as it is a very important part for choosing the price oracle in the first place. Leaving it here will help people to understand the rationale for the price oracle. 

If the sentence is confusing, please tell me how it can be improved!

Reply:
I consider this conversation as resolved because the sentence is rewritten (using correct jargon now). Furthermore, I think you are not anymore standing behind "I can't see any relationship of bins with pricing" (tell me if that is wrong!).

As we agreed on: we are merging this SWIP now. Feel free to raise this as a point of discussion with the wider community.